rllib/examples/checkpoints/change_config_during_training.py

```diff
@@ -0,0 +1,236 @@
+    LEARNER_RESULTS,
+)
+from ray.rllib.utils.numpy import convert_to_numpy
+from ray.rllib.utils.test_utils import (
+    add_rllib_example_script_args,
+    check,
+    run_rllib_example_script_experiment,
+)
+from ray.tune.registry import register_env
+
+
+parser = add_rllib_example_script_args(
+    default_reward=450.0, default_timesteps=10000000, default_iters=2000
+)
+parser.add_argument(
+    "--stop-reward-first-config",
+    type=float,
+    default=150.0,
+    help="Mean episode return after which the Algorithm on the first config should "
+    "stop training.",
+)
+# By default, set `args.checkpoint_freq` to 1 and `args.checkpoint_at_end` to True.
+parser.set_defaults(
+    enable_new_api_stack=True,
+    checkpoint_freq=1,
+    checkpoint_at_end=True,
+)
+
+
+if __name__ == "__main__":
+    args = parser.parse_args()
+
+    register_env(
+        "ma_cart", lambda cfg: MultiAgentCartPole({"num_agents": args.num_agents})
+    )
+
+    # Simple generic config.
+    base_config = (
+        PPOConfig()
+        .environment("CartPole-v1" if args.num_agents == 0 else "ma_cart")
+        .training(lr=0.0001)
+    )
+
+    # Setup multi-agent, if required.
+    if args.num_agents > 0:
+        base_config.multi_agent(
+            policies={
+                f"p{aid}": PolicySpec(
+                    config=AlgorithmConfig.overrides(
+                        lr=5e-5
+                        * (aid + 1),  # agent 1 has double the learning rate as 0.
+                    )
+                )
+                for aid in range(args.num_agents)
+            },
+            policy_mapping_fn=lambda aid, *a, **kw: f"p{aid}",
+        )
+
+    # Define some stopping criterion. Note that this criterion is an avg episode return
+    # to be reached.
+    metric = f"{ENV_RUNNER_RESULTS}/{EPISODE_RETURN_MEAN}"
+    stop = {metric: args.stop_reward_first_config}
+
+    tuner_results = run_rllib_example_script_experiment(
+        base_config,
+        args,
+        stop=stop,
+        keep_ray_up=True,
+    )
+
+    # Perform a very quick test to make sure our algo (upon restoration) did not lose
+    # its ability to perform well in the env.
+    # - Extract the best checkpoint.
+    best_result = tuner_results.get_best_result(metric=metric, mode="max")
+    assert (
+        best_result.metrics[ENV_RUNNER_RESULTS][EPISODE_RETURN_MEAN]
+        >= args.stop_reward_first_config
+    )
+    best_checkpoint_path = best_result.checkpoint.path
+
+    # Rebuild the algorithm (just for testing purposes).
+    test_algo = base_config.build()
+    # Load algo's state from best checkpoint.
+    test_algo.restore_from_path(best_checkpoint_path)
+    # Perform some checks on the restored state.
+    assert test_algo.training_iteration > 0
+    # Evaluate on the restored algorithm.
+    test_eval_results = test_algo.evaluate()
+    assert (
+        test_eval_results[ENV_RUNNER_RESULTS][EPISODE_RETURN_MEAN]
+        >= args.stop_reward_first_config
```

Comment:
Dumb question: Does this inequality always hold? We train for mean rewards, so theoretically there might be cases where this inequality does not hold, aren't there?

Reply:
True, but statistically, I think the chance is super low. We also *evaluate*, meaning we use greedy actions, which perform much stronger that the stochastic ones used during training.